### PR TITLE
feat(`forge`) - Test scaffolding

### PR DESCRIPTION
**Pull request by ravindranrahul**
> [!NOTE]  
> I have exactly 3 days of Rust experience. So I had to wing this PR. Please excuse if I have missed any obvious expectations. BTW rust blown away by the rust compiler :exploding_head: 

## Motivation
Creating test file typically involves repeated, verbose steps. This PR introduces a new forge sub command to help scaffold generation of test file - one of many features discussed in detail in this issue: #5466 

## Solution
A new forge sub command has been introduced `forge generate test --contract-name <CONTRACT_NAME>`. For a given contract it does the following:
- Creates the test file in `tests` directory.
- Adds the necessary imports.
- Adds setup code for the test contract.

A sub command called generate has intentionally been added to accommodate  more potential scaffolds in the future - like generating Interface for a given contract which also is a cumbersome exercise. 

cc: @mattsse @Evalir @0x-r4bbit

### CODE REVIEW

**Comment by ravindranrahul** [comment]
HI @Evalir - Let me know if anything else to be done. Thanks!

**Review by Evalir — APPROVED** [review]
looks good! supportive—pretty harmless addition. wdyt @mattsse ?

File: cli/assets/generated/TestTemplate.t.sol [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import "forge-std/Test.sol";
+import "../src/{contract_name}.sol";
```
Comment (mds1) [inline cli/assets/generated/TestTemplate.t.sol:5]: For consistency with the book and [best practices](https://book.getfoundry.sh/tutorials/best-practices), we should use named imports here
```solidity
import {Test, console2} from "forge-std/Test.sol";
import {ContractName} from "../src/{contract_name}.sol"; 
```

File: cli/assets/generated/TestTemplate.t.sol [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+pragma solidity ^0.8.13;
+
+import "forge-std/Test.sol";
+import "../src/{contract_name}.sol"; 
+
+contract {contract_name}Test is Test {
+    {contract_name} public {instance_name};
+
+    function setUp() public {
+        {instance_name} = new {contract_name}();
```
Comment (mds1) [inline cli/assets/generated/TestTemplate.t.sol:11]: Similarly, the best practices doc suggests scaffolding your tests by running your deploy script here, that way you're testing against the post-deploy state. Of course, not everyone does this, and it's not always applicable. But this is inconsistent with what's recommended, so wonder what the best approach is here. Perhaps we should add a comment linking to the best practices or something?

File: cli/assets/generated/TestTemplate.t.sol (hunk 1 above)
Comment (ravindranrahul) [reply cli/assets/generated/TestTemplate.t.sol:5]: I agree. Can we just import `Test` from `forge-std/Test.sol`? I believe console2 is not being used in the generated file.

File: cli/assets/generated/TestTemplate.t.sol (hunk 2 above)
Comment (ravindranrahul) [reply cli/assets/generated/TestTemplate.t.sol:11]: I have no comments on this. I will go with whatever the team here decides. As an aside, If we take any action, we may need to update the default `CounterTest` file too:

https://github.com/foundry-rs/foundry/blob/42beef0a1d6fc4a26a803e6321e041215123753e/cli/assets/CounterTemplate.t.sol#L8-L11

File: cli/assets/generated/TestTemplate.t.sol (hunk 1 above)
Comment (mds1) [reply cli/assets/generated/TestTemplate.t.sol:5]: > I believe `console2` is not being used in the generated file.

This is true, but it's convenient for users to have it imported by default. With `import "forge-std/Test.sol";` it's imported by default, but when you switch to named imports you have to include it explicitly

File: cli/assets/generated/TestTemplate.t.sol (hunk 2 above)
Comment (mds1) [reply cli/assets/generated/TestTemplate.t.sol:11]: I don't have a good generic solution here, so if we're in favor of adding this command we can leave this as-is. Will resolve this :)

File: cli/assets/generated/TestTemplate.t.sol (hunk 1 above)
Comment (ravindranrahul) [reply cli/assets/generated/TestTemplate.t.sol]: That makes sense. Thanks for reviewing Matt! I have resolved it.

File: cli/src/cmd/forge/generate/mod.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        let instance_name = format_identifier(&self.contract_name, false);
+
+        // Create the test file content.
+        let test_content = include_str!("../../../../assets/generated/TestTemplate.t.sol");
+        let test_content = test_content
+            .replace("{contract_name}", &contract_name)
+            .replace("{instance_name}", &instance_name);
+
+        // Create the test directory if it doesn't exist.
+        fs::create_dir_all("test").unwrap();
```
Comment (Evalir) [inline cli/src/cmd/forge/generate/mod.rs:38]: Could we avoid `unwrap` here? We can just use `?` and maybe `map_err` as well to make it a tad more descriptive.

File: cli/src/cmd/forge/generate/mod.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+        // Define the test file path
+        let test_file_path = Path::new("test").join(format!("{}.t.sol", contract_name));
+
+        // Write the test content to the test file.
+        fs::write(&test_file_path, test_content).unwrap();
```
Comment (Evalir) [inline cli/src/cmd/forge/generate/mod.rs:44]: ditto on unwrap

**Review by Evalir — CHANGES_REQUESTED** [review]
just some quick nits, but lgtm!

File: cli/src/cmd/forge/generate/mod.rs (hunk 3 above)
Comment (ravindranrahul) [reply cli/src/cmd/forge/generate/mod.rs:38]: Absolutely! How does the following code look?

```rust
    // Create the test directory if it doesn't exist
    fs::create_dir_all("test").map_err(|e| {
        format!(
            "Failed to create 'test' directory: {}. Please check if you have write access.",
            e
        )
    });
```

Comment (ravindranrahul) [reply cli/src/cmd/forge/generate/mod.rs]: Absolutely! I looked at some other files and figured there is wrapper around fs `foundry_common::fs` so I have swapped the implementation.

File: cli/src/cmd/forge/generate/mod.rs (hunk 4 above)
Comment (ravindranrahul) [reply cli/src/cmd/forge/generate/mod.rs]: fixed too

**Review by Evalir — COMMENTED** [review]
amazing, thank you!

**Review by Evalir — APPROVED** [review]

**Review by mattsse — APPROVED** [review]
this looks pretty good, nicely done

**Comment by mattsse** [comment]
that'd be a nice feature, please open a separate issue for this

**Pull request merged**